### PR TITLE
Ripple rewrite

### PR DESCRIPTION
**Pull request by robertmesserle**
@jelbourn @ThomasBurleson @rschmukler @topherfangio 

Please review these changes.  Please also check this out and look around for any breakages that I may have caused.

I will likely be going through and adding comments to make the code easier to navigate.

### CODE REVIEW

File: src/core/services/ripple/ripple.js [hunk 1]
```diff
@@ -1,382 +1,273 @@
...
-      element.on('$md.pressdown', onPressDown)
-        .on('$md.pressup', onPressUp);
-    }
-
-    controller.createRipple = createRipple;
+/**
+ * Controller used by the ripple service in order to apply ripples
+ * @ngInject
+ */
+function InkRippleServiceController ($scope, $element, rippleOptions, $window, $timeout, $mdUtil) {
```
Comment (jelbourn) [inline src/core/services/ripple/ripple.js]: I would just call this InkRippleCtrl

File: src/core/services/ripple/ripple.js [hunk 2]
```diff
@@ -1,382 +1,273 @@
...
+  this.$timeout   = $timeout;
+  this.$mdUtil    = $mdUtil;
+  this.$scope     = $scope;
+  this.$element   = $element;
+  this.options    = rippleOptions;
+  this.container  = this.createContainer();
+  this.mousedown  = false;
+  this.ripples    = [];
+  this.color      = this.getColor();
+  this.timeout    = null;
```
Comment (jelbourn) [inline src/core/services/ripple/ripple.js]: Comment for what `timeout` is?

File: src/core/services/ripple/ripple.js [hunk 3]
```diff
@@ -1,382 +1,273 @@
...
-        }
-        angular.forEach(ripples, updateElement);
-      });
-    }
+/**
+ * Returns the color that the ripple should be (either based on CSS or hard-coded)
+ * @returns {string}
+ */
+InkRippleServiceController.prototype.getColor = function () {
+  return parseColor(this.$element.attr('md-ink-ripple'))
```
Comment (jelbourn) [inline src/core/services/ripple/ripple.js]: Cache this value?

File: src/core/services/ripple/ripple.js [hunk 4]
```diff
@@ -1,382 +1,273 @@
...
+      || parseColor(getElementColor.call(this));
+
+  /**
+   * Finds the color element and returns its color
+   * @returns {string}
+   */
+  function getElementColor () {
+    var colorElement = this.options.colorElement && this.options.colorElement[ 0 ];
+    colorElement     = colorElement || this.$element[ 0 ];
+    return colorElement ? this.$window.getComputedStyle(colorElement).color : 'rgb(0,0,0)';
```
Comment (jelbourn) [inline src/core/services/ripple/ripple.js]: `rgba(0,0,0` > `black` ?

File: src/core/services/ripple/ripple.js [hunk 5]
```diff
@@ -1,382 +1,273 @@
...
+/**
+ * Returns the color that the ripple should be (either based on CSS or hard-coded)
+ * @returns {string}
+ */
+InkRippleServiceController.prototype.getColor = function () {
+  return parseColor(this.$element.attr('md-ink-ripple'))
+      || parseColor(getElementColor.call(this));
+
+  /**
+   * Finds the color element and returns its color
```
Comment (jelbourn) [inline src/core/services/ripple/ripple.js]: Add explanation of why you'd need to get the color of an element?

File: src/core/services/ripple/ripple.js [hunk 6]
```diff
@@ -1,382 +1,273 @@
...
-       * Converts a hex value to an rgba string
-       *
-       * @param {string} hex value (3 or 6 digits) to be converted
-       *
-       * @returns {string} rgba color with 0.1 alpha
-       */
-      function hexToRGBA(color) {
-        var hex = color.charAt(0) === '#' ? color.substr(1) : color,
+    function hexToRGBA (color) {
+      var hex = color.charAt(0) === '#' ? color.substr(1) : color,
```
Comment (jelbourn) [inline src/core/services/ripple/ripple.js]: color[0] should work in all browsers now

File: src/core/services/ripple/ripple.js [hunk 7]
```diff
@@ -1,382 +1,273 @@
...
+InkRippleServiceController.prototype.removeRipple = function (ripple) {
+  var ctrl  = this;
+  var index = this.ripples.indexOf(ripple);
+  if (index < 0) return;
+  this.ripples.splice(this.ripples.indexOf(ripple), 1);
+  ripple.removeClass('md-ripple-active');
+  if (this.ripples.length === 0) this.container.css({ backgroundColor: '' });
+  // use a 2-second timeout in order to allow for the animation to finish
+  // we don't actually care how long the animation takes
+  this.$timeout(function () { ctrl.fadeOutComplete(ripple); }, 650, false);
```
Comment (jelbourn) [inline src/core/services/ripple/ripple.js]: Extract specific timeouts into constants?

File: src/components/button/button.scss [hunk 8]
```diff
@@ -149,6 +154,14 @@ $icon-button-margin: rem(0.600) !default;
       }
     }
   }
+
+  .md-ripple-container {
+    border-radius: $button-border-radius;
+    background-clip: padding-box;
+    overflow: hidden;
+    // The following hack causes Safari/Chrome to respect overflow hidden for ripples
+    -webkit-mask-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAIAAACQd1PeAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAAA5JREFUeNpiYGBgAAgwAAAEAAGbA+oJAAAAAElFTkSuQmCC');
```
Comment (topherfangio) [inline src/components/button/button.scss]: Any chance we have a link to an article of why this works? Would be nice to add to the comment if available.

File: src/core/services/ripple/ripple.js [hunk 9]
```diff
@@ -1,382 +1,277 @@
...
-        controller = element.controller('mdInkRipple') || {},
-        counter = 0,
-        ripples = [],
-        states = [],
-        isActiveExpr = element.attr('md-highlight'),
-        isActive = false,
-        isHeld = false,
-        node = element[0],
-        rippleSizeSetting = element.attr('md-ripple-size'),
-        color = parseColor(element.attr('md-ink-ripple')) || parseColor(options.colorElement.length && $window.getComputedStyle(options.colorElement[0]).color || 'rgb(0, 0, 0)');
```
Comment (topherfangio) [inline src/core/services/ripple/ripple.js]: This line feels a bit long and could either use a comment or a refactoring to make it a bit more clear.

File: src/core/services/ripple/ripple.js [hunk 10]
```diff
@@ -1,382 +1,277 @@
...
-            width: rippleSize + 'px',
-            height: rippleSize + 'px',
-            marginLeft: (rippleSize * -1) + 'px',
-            marginTop: (rippleSize * -1) + 'px'
-          });
-        }
-        removeElement(elem, options.outline ? 450 : 650);
+      if (dig === 1) {
+        red += red;
+        grn += grn;
```
Comment (topherfangio) [inline src/core/services/ripple/ripple.js]: Any particular reason this is shortened to `grn` and `blu`?

File: src/core/services/ripple/ripple.js [hunk 11]
```diff
@@ -1,382 +1,282 @@
...
+  this.$timeout   = $timeout;
+  this.$mdUtil    = $mdUtil;
+  this.$scope     = $scope;
+  this.$element   = $element;
+  this.options    = rippleOptions;
+  this.container  = this.createContainer();
+  this.mousedown  = false;
+  this.ripples    = [];
+  this.color      = this.getColor();
+  this.background = this.getColor(0.5);
```
Comment (ThomasBurleson) [inline src/core/services/ripple/ripple.js]: These should be deferred until the ripple render occurs (and then cached). See https://github.com/angular/material/issues/4199 for details.

**Pull request closed without merging**